### DISCUSSION TITLE
Improving numerical stability

### DISCUSSION THREAD

**Discussion by OrtnerMichael** [magpylib/magpylib#665] [General]
A lot of our problems are related with numerical stability - it has accompanied us for many years and can be tracked through multiple issues in the past

- #648
- #643 
- #547 
- #467 
- #404
- #374 
- #370 
- #242 
- #241 

We already spent a lot of time on this - also there are a few papers that are dealing with this issue

https://www.mdpi.com/2673-8724/3/1/2

https://ieeexplore.ieee.org/abstract/document/8603829

https://onlinelibrary.wiley.com/doi/full/10.1002/advs.202301033

It's time that we payed more attention to this issue ! Specifically we should solve two problems:
1. The cylinder stability, as it is proposed in Masiero2023, and based on PeterLeiters unpublished work
2. The Cuboid stability, review the Cichon2019 solution and discuss the long-distance problem - possibly by using the dipole solution

both of these issues are straightforward - they just need to be worked out properly, analyzed, implemented and tested.

@peter-leitner how far are you with the Cylinder stability ?

**Comment by fslanovc** [comment]
If you sweep to infinity along one of the space diagonals, the cube field becomes unstable at a certain point because it is numerically cancelled out

![Figure_1](https://github.com/magpylib/magpylib/assets/52533009/ec60d985-93cd-42c7-a676-fe052c1648bd)

**Reply by OrtnerMichael** [reply]
Thank you @fslanovc for this plot.

Concerning stability in the near-field: In Cichons paper it is written that the Cuboid field is stable in the first ocant only. This is also how Magpylib is implemented - but I never managed to understand what is really the origin of the instability.

Transforming everything to the first octant is done with numpy masks - not very efficient. If we could reformulate the expressions so that all octants are stable, this would be a huge improvement.

Also, in Cichons paper he only states that there is an instability, but he never analyzes it in detail. To me it seems that such an Analysis might have also a scientific value in terms of publication.

**Comment by OrtnerMichael** [comment]
@peter-leitner 
Can you please give us a status update on the Cylinder by replying to this post ?

**Comment by peter-leitner** [comment]
The homogenously magnetized cylinder (both, the longitudinal and diametral contributions) analytical field solutions as presented e.g. in [Caciagli et al. 2018, Journal of Magnetism and Magnetic Materials 456:423-423] show inherent numerical instabilities when evaluating the field either very close to the symmetry axis as well as in the far field. The field solutions, typically stated in a form combining elliptical integrals of the first, second and third kind completely lose their numerical accuracy to the degree even of zero-precision floating point numbers due to catastrophic cancellation of these additive terms in the aforementioned fieldpoint regions of R³. The loss of precision is not an academic concern as the analytic formulae are correct in their own regard but rather affect specific applications based on analytics where the unaware practicioner might encounter unexpected behaviour when relying on these expressions. Such applications might be any sort of magnetic positioning systems whenever sensor measurements are compared to analytically determined field values for magnetostatic inversion. One might think in particular about end-of-shaft angle-sensing systems, where the field sensors are placed directly below cylindrical magnets co-rotating on a shaft, where the corresponding analytic field evaluation will degrade the closer the cylinder axis approaches the sensor.

In order to address this problem in Magpylib, we worked on the stabilization of the problematic analytic expressions and made use of the fact that linear combinations of complete Legendre integrals can be expressed through the so-called generalized complete Bulirsch integral. While the cel-algorithm as can be found in the literature is not numerically stable throughout by itself, a specifically designated version for the current-loop field application has been implemented by [Ortner et al. 2023, MDPI Magnetism 3:11-31]. The Bulirsch algorithm significantly improves the numerical instability issue in particular throughout the typical ranges of interest in applications by adding up to 8 significant figures compared to the straight-forward implementation of the Caciagli formulas, however not the entire numerical instability problem can be removed.

Besides substituting the field formulas by series expansions and dipole approximation in the problematic regions, we are currently also working on improving the stability of the Bulirsch algorithm for the diametral cylinder problem similar to the work done by [Ortner et al. 2023].

![accuracy_plot_Caciagli_3x1](https://github.com/magpylib/magpylib/assets/51168238/92eb0350-4ccd-4fca-b4db-30fff5956ac1)
Loss of numerical precision in the analytical expressions derived by [Caciagli et al. 2018] shown for a range typically needed in applications.

![cel-solution](https://github.com/magpylib/magpylib/assets/51168238/378e5e8a-04a3-463d-af35-0830aad74601)
Analytical solutions stabilized by the Bulirsch algorithm. As can be seen, the numerical instability recedes further out and higher precision can be achieved in the boxed area (same range as in the figure above).

![grafik](https://github.com/magpylib/magpylib/assets/51168238/7cdf6e58-7ad9-4db1-bea3-d3fc016c111d)
A series expansion (shown here for the field around the cylinder axis) can give an even higher accuracy in the problematic area it is supposed to address.

![grafik](https://github.com/magpylib/magpylib/assets/51168238/73fd3cda-0ebf-4d4f-91f9-b55a69b8bf1b)
Proper combinations of series expansions and the analytic field formulas can give a decent numerical precision of at least eight digits throughout (here shown for a typical vertical distance of 1 cm from the magnet for small to medium radii.
